### PR TITLE
Support overriding virtual indexer property in proxies

### LINKED ISSUE #21985: Change-tracking proxies don't detect changes in Indexer properties
See https://github.com/dotnet/efcore/pull/21948#discussion_r466571755

### PR DESCRIPTION
**Pull request by ajcvickers**
Fixes #21985

### CODE REVIEW

**Comment by ajcvickers** [comment]
@Pilchie RC1

**Comment by smitpatel** [comment]
May be should also remove line from https://github.com/dotnet/efcore/pull/21948#discussion_r466113532

**Comment by ajcvickers** [comment]
@smitpatel Dictionary is not proxyable. We might want to throw if people try to use it as a proxy type, but that would also fail all our current tests. We should discuss in the context of #22009.

**Comment by Pilchie** [comment]
Approved for RC1.

**Comment by ajcvickers** [comment]
@smitpatel @AndriySvyryd Added another commit which improves this significantly by handling null values returned from indexer properties in the same way we handle null values in the backing field of a non-nullable property.

File: src/EFCore/ChangeTracking/Internal/SnapshotFactoryFactory.cs [hunk 1]
```diff
@@ -145,7 +145,12 @@ public virtual Func<ISnapshot> CreateEmpty([NotNull] IEntityType entityType)
 
                 if (memberAccess.Type != propertyBase.ClrType)
                 {
-                    memberAccess = Expression.Convert(memberAccess, propertyBase.ClrType);
+                    var hasDefaultValueExpression = memberAccess.MakeHasDefaultValue(propertyBase);
+
+                    memberAccess = Expression.Condition(
+                            hasDefaultValueExpression,
+                            propertyBase.ClrType.GetDefaultValueConstant(),
+                            Expression.Convert(memberAccess, propertyBase.ClrType));
```
Comment (smitpatel) [inline src/EFCore/ChangeTracking/Internal/SnapshotFactoryFactory.cs:153]: Probably 1 extra indent.

File: src/EFCore/Extensions/Internal/ExpressionExtensions.cs [hunk 2]
```diff
@@ -29,9 +29,9 @@ public static class ExpressionExtensions
         ///     any release. You should only use it directly in your code with extreme caution and knowing that
         ///     doing so can result in application failures when updating to a new Entity Framework Core release.
         /// </summary>
-        public static Expression MakeHasDefaultValue<TProperty>(
+        public static Expression MakeHasDefaultValue(
             [NotNull] this Expression currentValueExpression,
-            [NotNull] IPropertyBase propertyBase)
+            [CanBeNull] IPropertyBase propertyBase)
```
Comment (smitpatel) [inline src/EFCore/Extensions/Internal/ExpressionExtensions.cs:34]: What is passing null here now?

Comment (ajcvickers) [reply src/EFCore/Extensions/Internal/ExpressionExtensions.cs:34]: We use this to create DbSet property setters. That's where.

Comment (smitpatel) [reply src/EFCore/Extensions/Internal/ExpressionExtensions.cs:34]: Ah so annotation was incorrect before this PR itself. I was looking for change in this PR which was passing null.

Comment (ajcvickers) [reply src/EFCore/Extensions/Internal/ExpressionExtensions.cs:34]: No, it wasn't being called in that path before. Tests failed.

**Pull request merged**